Login: Add send me a login link to the password validation message

StefanNieuwenhuis opened the pull request:
Note from @escapemanuele on 2023-10-19: We will work on this PR to get it up to date, and check if it is still useful.


### Changes proposed in this Pull Request

Adding a login link to the password validation message to improve the user experience for the login page.

#### Translator notes
> Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?

When a user provides the wrong password, a validation message is shown. The to-be-translated copy is part of it (see screenshot).

<img width="547" alt="Screenshot 2021-07-13 at 11 05 05" src="https://user-images.githubusercontent.com/12104589/125424423-798ce33d-8dbe-44bd-80f2-1ed5b552fbca.png">


#### Technical changes
* `canDoMagicLogin`: Util to validate if the magic login link is allowed to be sent.
* `getLoginLinkPageUrl`: Util to generate a magic login link.
* `renderMagicLoginLink`: Method to render the magic login link in the validation message.

### Testing instructions
1. Check out this branch and `yarn && yarn start`.
2. Apply D62486-code to your sandbox.
3. Make sure that public-api.wordpress.com is proxied.
4. Go to `/log-in`
5. Enter an existing username/email address and click on "continue".
6. Enter a wrong password deliberately.
7. Confirm that:
    * [ ] The validation message copy is updated to _"Oops, that's not the right password. Would you like to receive a login link?"_
    * [ ] Clicking on the link redirects to the _email-me-a-login-link_ view.

Related to #53379

StefanNieuwenhuis committed:
Initial commit

StefanNieuwenhuis committed:
Add utils for magic login links

* `getLoginPageLinkUrl`: Util to generate a magic login link.
* `canDoMagicLogin`: Util to validate if magic logins are allowed.

StefanNieuwenhuis committed:
Update pwd validation msg

Use utils to validate magic login and generate a magic login link

alshakero reviewed (commented):
Can you rebase this, so I could test it against the newest codebase please?

StefanNieuwenhuis commented:
> Can you rebase this, so I could test it against the newest codebase please?

Done! Can you please give it another try?

alshakero commented on client/blocks/login/login-form.jsx:
```diff
@@ -448,6 +448,36 @@ export class LoginForm extends Component {
+		}
+
+		return this.props.translate(
+			' Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?',
```
```suggestion
			'Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?',
```

It's also not clear where the user will receive the link. I would add "to your email".

StefanNieuwenhuis replied:
> It's also not clear where the user will receive the link. I would add "to your email".

The user is redirected to `/wp-login.php?action=lostpassword` where the user gets more instructions about the next steps (see screenshot). So adding "to your email" might be a bit overkill? 

@ollierozdarz, WDYT?


<img width="547" alt="Screenshot 2021-07-13 at 09 40 09" src="https://user-images.githubusercontent.com/12104589/125411564-538d6980-e3be-11eb-8063-7b4b84895185.png">

StefanNieuwenhuis replied:
Although it looks like a typo, the leading space is necessary since the first part of the validation message comes from the BE and has no trailing space. I added a comment in b8233a69da936d2128dc387a9ec887c131a1569f

alshakero replied:
I see. But I'd add the space outside of the translation. It's all optional though.

alshakero reviewed (approved):
Very neatly implemented, tested and works as expected. Left one comment.

razvanpapadopol commented on client/lib/login/index.js:
```diff
@@ -112,3 +113,44 @@ export function getSignupUrl(
 	return signupUrl;
 }
+
+export const getLoginLinkPageUrl = ( locale, currentRoute, isGutenboarding ) => {
+	const loginParameters = {
+		locale,
+		twoFactorAuthType: 'link',
+	};
+
+	if ( currentRoute === '/login-in/jetpack' ) {
+		loginParameters.twoFactorAuthType = 'jetpack/link';
+	} else if ( isGutenboarding ) {
+		loginParameters.twoFactorAuthType = 'new/link';
+	}
+
+	return login( loginParameters );
+};
```
Good to see this piece of logic extracted to lib. Shouldn't we re-use it also in `LoginLinks` component? https://github.com/Automattic/wp-calypso/blob/update/login-form%2Fwrong-pwd-validation-msg/client/login/wp-login/login-links.jsx#L104-L121

StefanNieuwenhuis replied:
Great suggestion! I've added it in 53138a563f81928820b1d81931a4e81e2d65a6f0

razvanpapadopol commented on client/lib/login/index.js:
```diff
@@ -112,3 +113,44 @@ export function getSignupUrl(
+	return login( loginParameters );
+};
+
+export const canDoMagicLogin = (
+	twoFactorAuthType,
+	oauth2Client,
+	wccomFrom,
+	isJetpackWooCommerceFlow
+) => {
+	if ( ! config.isEnabled( `login/magic-login` ) || twoFactorAuthType ) {
+		return false;
+	}
+
+	// jetpack cloud cannot have users being sent to WordPress.com
+	if ( isJetpackCloudOAuth2Client( oauth2Client ) ) {
+		return false;
+	}
+
+	if ( isWooOAuth2Client( oauth2Client ) && wccomFrom ) {
+		return false;
+	}
+
+	if ( isJetpackWooCommerceFlow ) {
+		return false;
+	}
+
+	return true;
+};
```
Same as above https://github.com/Automattic/wp-calypso/pull/53499#discussion_r665457321
Related code that should re-use this logic: https://github.com/Automattic/wp-calypso/blob/update/login-form%2Fwrong-pwd-validation-msg/client/login/wp-login/login-links.jsx#L209-L228

StefanNieuwenhuis replied:
Added in 53138a563f81928820b1d81931a4e81e2d65a6f0

razvanpapadopol reviewed (changes requested):
This is such a nice improvement to the Login experience! Thanks, @StefanNieuwenhuis!

I've added the "String Freeze" label since we're introducing new copy.

Then I've left two notes on possible re-usability. Usually this kind of stuff is not blocking but in this case I guess we can do it while we're waiting for the translations.

Finally, extracting this logic to pure functions may be an opportunity for us to add some unit tests. I see that other `client/lib` folders have test coverage so it would be good for `client/lib/login` to join the party 🎉 
In the same spirit, I'd consider updating also the unit tests covering the [LoginForm component](https://github.com/Automattic/wp-calypso/blob/1d2278647dc980ae462ef6a472464d628d4a6930/client/blocks/login/test/login-form.jsx) if possible.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6153525

Thank you @StefanNieuwenhuis for including a screenshot in the description! This is really helpful for our translators.

StefanNieuwenhuis committed:
Update login links

Make use of the getLoginLinkPageUrl from client/lib/login to make the code more generic ad robust.

StefanNieuwenhuis committed:
Remove redudant leading space in magic login link msg

a8ci18n commented on client/blocks/login/login-form.jsx:
```diff
@@ -448,6 +448,36 @@ export class LoginForm extends Component {
+		}
+
+		return this.props.translate(
+			'Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **14** times:
`translate( ' Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?' )` *ES Score: 13*

StefanNieuwenhuis committed:
Update rendering logic for magic login link

StefanNieuwenhuis committed:
Remove debug string

StefanNieuwenhuis committed:
Restore leading space and add comment

a8ci18n commented on client/blocks/login/login-form.jsx:
```diff
@@ -448,6 +448,37 @@ export class LoginForm extends Component {
+
+		// The leading space is NO typo! It's added since the first part of the validation message (originating from the BE) contains no trailing space.
+		return this.props.translate(
+			' Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?',
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **14** times:
`translate( 'Would you like to {{magicLoginLink}}receive a login link{{/magicLoginLink}}?' )` *ES Score: 16*

StefanNieuwenhuis committed:
Add unit tests for login lib

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6153525).

StefanNieuwenhuis commented:
> This is such a nice improvement to the Login experience! Thanks, @StefanNieuwenhuis!
> 
> I've added the "String Freeze" label since we're introducing new copy.
> 
> Then I've left two notes on possible re-usability. Usually this kind of stuff is not blocking but in this case I guess we can do it while we're waiting for the translations.
> 
> Finally, extracting this logic to pure functions may be an opportunity for us to add some unit tests. I see that other `client/lib` folders have test coverage so it would be good for `client/lib/login` to join the party 🎉
> In the same spirit, I'd consider updating also the unit tests covering the [LoginForm component](https://github.com/Automattic/wp-calypso/blob/1d2278647dc980ae462ef6a472464d628d4a6930/client/blocks/login/test/login-form.jsx) if possible.

Thanks for your review and suggestions @razvanpapadopol! I've added unit tests for the newly created lib in abd6876206686f044f2cc14ab4e8c5304501acb0 but purposely left out the integration tests for `<LoginForm/>` since they will bloat this PR. I created a new issue (https://github.com/Automattic/wp-calypso/issues/55023) to address them.

alshakero commented on client/lib/login/test/index.js:
```diff
@@ -0,0 +1,171 @@
+	getLoginLinkPageUrl,
+	getSignupUrl,
+	getSocialServiceFromClientId,
+} from '..';
```
Optional: 
```suggestion
} from 'calypso/lib/login';
```

alshakero reviewed (approved):
Great work! Left optional notes.

StefanNieuwenhuis committed:
Update client/lib/login/test/index.js

Change import path to something that makes much more sense.

Co-authored-by: Omar Alshaker <omar@omaralshaker.com>

StefanNieuwenhuis commented:
> I see. But I'd add the space outside of the translation. It's all optional though.

Thanks so much @alshakero!
I leave it in for now since otherwise, it'll trigger a new translation cycle. 
I implemented your other suggestion for the import. Great find, which makes the code much more readable.

StefanNieuwenhuis committed:
Fix merge conflicts

razvanpapadopol commented:
Thanks! The only thing left here is to approve D62486-code

escapemanuele commented:
Done here: https://github.com/Automattic/wp-calypso/pull/83435
